src/carcockpit/gltf_viewer_widget.hpp

```diff
@@ -0,0 +1,113 @@
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+/* ================ LICENSE END ================ */
+
+#pragma once
+
+#include <ruis/res/texture_2d.hpp>
+#include <ruis/res/texture_cube.hpp>
+#include <ruis/updateable.hpp>
+#include <ruis/widgets/base/fraction_widget.hpp>
+#include <ruis/widgets/widget.hpp>
+
+// #include "../ruis/render/scene/shaders/shader_adv.hpp"
+// #include "../ruis/render/scene/shaders/shader_phong.hpp"
+// #include "../ruis/render/scene/shaders/shader_skybox.hpp"
+#include "../ruis/render/scene/scene.hpp"
+#include "../ruis/render/scene/scene_renderer.hxx"
+
+#include "shaders/shader_adv.hpp"
+#include "shaders/shader_phong.hpp"
+#include "shaders/shader_skybox.hpp"
+
+namespace carcockpit {
+
+constexpr ruis::vec3 default_camera_position_front{0, 1.5, 7};
+constexpr ruis::vec3 default_camera_position_top{0, 9, 0};
+constexpr float default_camera_transition_duration{0.1};
+
+class gltf_viewer_widget : public ruis::fraction_widget, public ruis::updateable, virtual public ruis::widget
+{
+	std::shared_ptr<ruis::render::scene> demoscene;
+	std::shared_ptr<ruis::render::scene_renderer> sc_renderer;
+	std::shared_ptr<ruis::render::camera> camrip;
+	// std::shared_ptr<ruis::res::texture_cube> texture_environment_cube;
+
+	ruis::vec3 camera_position_front{default_camera_position_front};
+	ruis::vec3 camera_position_top{default_camera_position_top};
+	ruis::vec3 camera_position{camera_position_top};
+	ruis::vec3 camera_attractor{camera_position_front};
+
+	ruis::real camera_transition_duration{default_camera_transition_duration}; // not seconds
+	bool mouse_orbit = false;
+	// bool mouse_pan = false;
+	ruis::vec2 mouse_changeview_start;
+	ruis::vec3 camera_changeview_start;
+
+	uint32_t fps = 0;
+	uint32_t fps_sec_counter = 0;
+	uint32_t time = 0;
+
+public:
+	struct parameters {
+		std::string path_to_gltf;
+		float scaling_factor{1.0f};
+		ruis::vec3 camera_target{0, 0, 0};
+		bool smooth_navigation_orbit = true;
+		bool smooth_navigation_zoom = true;
+		ruis::real orbit_angle_upper_limit = M_PI_2;
+		ruis::real orbit_angle_lower_limit = M_PI_2;
+		std::shared_ptr<ruis::res::texture_cube> environment_cube;
+	};
+
+private:
+	parameters params;
+
+public:
+	struct all_parameters {
+		ruis::widget::parameters widget_params;
+		parameters gltf_params;
+	};
+
+	gltf_viewer_widget(utki::shared_ref<ruis::context> context, all_parameters params);
+
+	void render(const ruis::matrix4& matrix) const override;
+	void update(uint32_t dt) override;
+	void toggle_camera(bool toggle);
```

Comment:
add comment describing what does it toggle about camera?

Reply:
Since it is a part of public api, the comment should be a doxygen comemnt:
```
/**
 * @brief Brief description.
 * Optional detailed description.
 * @param p - param description.
 * ...
 * @return Optional return value description.
 */
```

Reply:
I removed this method and all corresponding fields, because it was made for the car viewer specifically, and is not relevant for a more common case like an arbitrary 3d model viewer